src/BenchmarkDotNet.Diagnostics.Windows/EtwDiagnoser.cs

```diff
@@ -1,18 +1,103 @@
-﻿using BenchmarkDotNet.Parameters;
+﻿using System;
+using System.Collections.Concurrent;
+using BenchmarkDotNet.Parameters;
 using BenchmarkDotNet.Running;
 using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Diagnostics.Tracing;
+using Microsoft.Diagnostics.Tracing.Parsers;
+using Microsoft.Diagnostics.Tracing.Session;
 
 namespace BenchmarkDotNet.Diagnostics.Windows
 {
-    public abstract class ETWDiagnoser
+    public abstract class EtwDiagnoser<TStats> where TStats : new()
     {
+        protected readonly LogCapture Logger = new LogCapture();
         protected readonly List<int> ProcessIdsUsedInRuns = new List<int>();
+        protected readonly ConcurrentDictionary<int, TStats> StatsPerProcess = new ConcurrentDictionary<int, TStats>();
 
-        protected string GetSessionName(string prefix, Benchmark benchmark, ParameterInstances parameters = null)
+        private TraceEventSession session;
+
+        protected abstract ClrTraceEventParser.Keywords EventType { get; }
+
+        protected abstract string SessionNamePrefix { get; }
+
+        protected void Start(Process process, Benchmark benchmark)
+        {
+            Cleanup();
+
+            ProcessIdsUsedInRuns.Add(process.Id);
+            StatsPerProcess.TryAdd(process.Id, new TStats());
+
+            AttachToEvents(CreateSession(benchmark), benchmark);
+
+            EnableProvider();
+
+            // The ETW collection thread starts receiving events immediately, but we only
+            // start aggregating them after ProcessStarted is called and we know which process
+            // (or processes) we should be monitoring. Communication between the benchmark thread
+            // and the ETW collection thread is through the statsPerProcess concurrent dictionary
+            // and through the TraceEventSession class, which is thread-safe.
+            var task = Task.Factory.StartNew((Action)(() => session.Source.Process()), TaskCreationOptions.LongRunning);
+
+            // wait until the processing has started, block by then so we don't loose any 
+            // information (very important for jit-related things)
+            WaitUntilStarted(task); 
+        }
+
+        protected void Stop()
+        {
+            // ETW real-time sessions receive events with a slight delay. Typically it
+            // shouldn't be more than a few seconds. This increases the likelihood that
+            // all relevant events are processed by the collection thread by the time we
+            // are done with the benchmark.
+            Thread.Sleep(TimeSpan.FromSeconds(3));
```

Comment:
Could we wrap this line in a separated method with nice name (like `WaitFor...`)?